qcodes/utils/logger.py

```diff
@@ -0,0 +1,439 @@
+import io
+import logging
+
+import os
+from pathlib import Path
+from collections import OrderedDict
+from contextlib import contextmanager
+from copy import copy
+
+from typing import Optional, List, Union, Sequence
+
+import pandas
+from pandas.core.series import Series
+
+import qcodes as qc
+from qcodes.instrument.base import InstrumentFilter, InstrumentBase
+
+LevelType = Union[int, str]
+
+
+log = logging.getLogger(__name__)
+
+LOGGING_DIR = "logs"
+LOGGING_SEPARATOR = ' ¦ '
+HISTORY_LOG_NAME = "command_history.log"
+PYTHON_LOG_NAME = 'qcodes.log'
+QCODES_USER_PATH_ENV = 'QCODES_USER_PATH'
+
+FORMAT_STRING_DICT = OrderedDict([
+    ('asctime', 's'),
+    ('name', 's'),
+    ('levelname', 's'),
+    ('funcName', 's'),
+    ('lineno', 'd'),
+    ('message', 's')])
+FORMAT_STRING_ITEMS = [f'%({name}){fmt}'
+                       for name, fmt in FORMAT_STRING_DICT.items()]
+
+# The handler of the root logger that get set up by `start_logger` are globals
+# for this modules scope, as it is intended to only use a single file and
+# console hander.
+console_handler: Optional[logging.Handler] = None
+file_handler: Optional[logging.Handler] = None
+
+
+def get_console_handler() -> Optional[logging.Handler]:
+    """
+    Get h that prints messages from the root logger to the console.
+    Returns `None` if `start_logger` had not been called.
+    """
+    global console_handler
+    return console_handler
+
+
+def get_file_handler() -> Optional[logging.Handler]:
+    """
+    Get h that streams messages from the root logger to the qcdoes log
+    file. To setup call `start_logger`.
+    Returns `None` if `start_logger` had not been called
+    """
+    global file_handler
+    return file_handler
+
+
+def get_level_name(level: Union[str, int]) -> str:
+    if isinstance(level, str):
+        return level
+    elif isinstance(level, int):
+        return logging.getLevelName(level)
+    else:
+        raise RuntimeError('get_level_name: '
+                           f'Cannot to convert level {level} of type '
+                           f'{type(level)} to logging level name. Need '
+                           'string or int.')
+
+def get_level_code(level: Union[str, int]) -> int:
+    if isinstance(level, int):
+        return level
+    elif isinstance(level, str):
+        return logging.getLevelName(level)
+    else:
+        raise RuntimeError('get_level_code: '
+                           f'Cannot to convert level {level} of type '
+                           f'{type(level)} to logging level code. Need '
+                           'string or int.')
+
+def _get_qcodes_user_path() -> str:
+    """
+    Get '~/.qcodes' path or if defined the path defined in the QCODES_USER_PATH
```

Comment:
This is hard coded because we want to have uniformity across fridges.

Reply:
Do you have any dependencies on the logs except users opening them in an editor?
The reason for the wish for uniformity is that there should be another process, that is independent of qcodes, to read the log files. You could argue that it could also readout this location from the qcodes config. But it seems to be a little to much that it would need to implement qcodes or go through the same cascading loading of configs without qcodes.

Reply:
you're right that qcodes config is indeed not a good location for this info.

I am just striving for flexibility, so that, if necessary, a painless change is always possible. With these expectations in mind, I also hope that the process that reads the logs is configurable, hence I can tell it where to read the logs from.

So, to conclude, it is good that there is an environment variable that allows to customize this location.